### PR TITLE
Squid 425 bug 4906 acl missmatches on tcp probes

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**

### CODE REVIEW

File: src/DelayId.cc [hunk 1]
```diff
@@ -90,22 +90,13 @@ DelayId::DelayClient(ClientHttpRequest * http, HttpReply *reply)
...
-        if (Config.onoff.delay_pool_uses_indirect_client)
-            ch.src_addr = r->indirect_client_addr;
-        else
-#endif /* FOLLOW_X_FORWARDED_FOR */
-            ch.src_addr = r->client_addr;
-        ch.my_addr = r->my_addr;
-
-        if (http->getConn() != NULL)
-            ch.conn(http->getConn());
+        ch.configureClientAddr(Config.onoff.delay_pool_uses_indirect_client);
```
Comment (rousskov) [inline src/DelayId.cc]: The Old code used `r->client_addr` when Config.onoff.delay_pool_uses_indirect_client was off, regardless of whether Config.onoff.acl_uses_indirect_client was set. The new code checks both configuration options. AFAICT, only the old behavior was correct. There is another change request that prompted the same question: Should we refactor to call forceIndirectAddr() here instead (when needed)?

File: src/HttpRequest.cc [hunk 2]
```diff
@@ -116,6 +115,19 @@ HttpRequest::init()
...
     forcedBodyContinuation = false;
+    internal = false;
+
+    if (clientConnectionManager().valid()) {
+        if (const auto port = clientConnectionManager()->port) {
+            flags.ignoreCc = port->ignore_cc;
+        }
+    }
+#if FOLLOW_X_FORWARDED_FOR
+    // indirect client gets stored here because it is an HTTP header result (from X-Forwarded-For:)
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/HttpRequest.cc]: I removed this code. HttpRequest::indirectClientAddr() method should properly deal with not-initialized indirect_client_addr field.

File: src/HttpRequest.cc [hunk 3]
```diff
@@ -802,9 +866,14 @@ FindListeningPortAddress(const HttpRequest *callerRequest, const AccessLogEntry
...
+    ip = FindListeningPortAddressInConn(request->clientConnection());
     if (!ip && ale)
         ip = FindListeningPortAddressInConn(ale->tcpClient);
     return ip; // may still be nil
 }
 
+Comm::ConnectionPointer
+HttpRequest::clientConnection() const
+{
+    return masterXaction->clientConnection();
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/HttpRequest.cc]: Done.

File: src/MasterXaction.h [hunk 4]
```diff
@@ -40,21 +43,35 @@ class MasterXaction : public RefCountable
 {
 public:
     typedef RefCount<MasterXaction> Pointer;
+    explicit MasterXaction(const XactionInitiator);
+
+    MasterXaction(const XactionInitiator, ConnStateData *);
+
+    MasterXaction(const XactionInitiator, Comm::ConnectionPointer);
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/MasterXaction.h]: > The idea is to let the user a choice of two options: pass either manager or connection, but not both.

I understand the rationale now, thank you. To be more precise, we are forcing a choice among three (not two) options: manager, connection, or nothing. I think this API is acceptable and possibly even the best (within this PR scope). Since those that poses ConnStateData can use two constructors, let's codify the rule:

```
    // Creators must call the first constructor they can call (with a non-nil
    // pointer). Following this rule maximizes stored information.
    MasterXaction(const XactionInitiator, ConnStateData *);
    MasterXaction(const XactionInitiator, Comm::ConnectionPointer);    
    explicit MasterXaction(const XactionInitiator anInitiator) : initiator(anInitiator) {};
```

File: src/tunnel.cc [hunk 5]
```diff
@@ -1101,15 +1101,13 @@ tunnelStart(ClientHttpRequest * http)
      * be allowed.  yuck, I know.
      */
 
-    if (Config.accessList.miss && !request->client_addr.isNoAddr()) {
+    if (Config.accessList.miss && !request->clientAddr().isNoAddr()) {
```
Comment (rousskov) [inline src/tunnel.cc]: Should we use the new `request->internal` here (and remove or adjust the above comment)? Or are there cases where internal (in the comment's sense) requests do not have `request->internal` set? Please double check whether additional polishing is needed to keep `request->internal` in sync with `request->clientAddr().isNoAddr()` and adjust the above source code comment as needed.

File: src/HttpRequest.h [hunk 6]
```diff
@@ -251,6 +264,16 @@ class HttpRequest: public Http::Message
     /// annotations added by the note directive and helpers
     /// and(or) by annotate_transaction/annotate_client ACLs.
     NotePairs::Pointer theNotes;
+
+    Ip::Address client_addr;
```
Comment (rousskov) [inline src/HttpRequest.h]:
```suggestion
    Ip::Address client_addr; ///< source address of a non-TCP (e.g. ICMP) client
```

And please also document `my_addr`.

File: src/HttpRequest.h [hunk 7]
```diff
@@ -251,6 +264,16 @@ class HttpRequest: public Http::Message
...
+    Ip::Address my_addr;
+#if FOLLOW_X_FORWARDED_FOR
+    Ip::Address indirect_client_addr;
```
Comment (rousskov) [inline src/HttpRequest.h]: Let's document this address if possible.

File: src/HttpRequest.cc [hunk 8]
```diff
@@ -714,34 +714,96 @@ UpdateRequestNotes(ConnStateData *csd, HttpRequest &request, NotePairs const &he
...
+        return NoAddr();
+    if (!indirect_client_addr.isEmpty())
+        return indirect_client_addr;
+    return clientAddr();
+}
+
+void
+HttpRequest::ignoreIndirectClientAddr()
+{
+    indirect_client_addr = clientAddr();
```
Comment (rousskov) [inline src/HttpRequest.cc]: The code does not seem to match the method name -- we are setting (a part of) indirect_client_addr instead of making it unusable/empty. Perhaps we can we use setEmpty() here instead?

File: src/acl/FilledChecklist.cc [hunk 9]
```diff
@@ -239,23 +247,81 @@ ACLFilledChecklist::ACLFilledChecklist(const acl_access *A, HttpRequest *http_re
...
+        my_addr = request->myAddr();
+        setClientConnectionManager(request->clientConnectionManager().get());
+        setClientConnection(request->clientConnection());
+    }
+}
 
-        if (request->clientConnectionManager.valid())
-            conn(request->clientConnectionManager.get());
+void
+ACLFilledChecklist::clientConnectionManager(ConnStateData *aConn)
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/acl/FilledChecklist.cc]: > The intention was rather simple: public methods ..., private methods ...

Just FYI: The problem with that approach is that the Squid code reader cannot easily tell the difference between public and private methods. At least I could not. Requiring the reader to memorize the public/private mapping is not a good option, of course.

File: src/tunnel.cc (hunk 5 above)
Comment (rousskov) [reply src/tunnel.cc]: > it looks like setNoAddr() is not called for source addresses for these internal requests and hence the existing check work incorrectly. Should we fix this inconsistency here, in this PR?

We probably should fix it (and it may be fixed automatically) _if_ we should modify this code to use our new HttpRequest::internal flag. We probably should fix it also if the fix is trivial and well-researched/understood (e.g., replacing isNoAddr() with a flag check). Otherwise, we probably should not fix this but may add a TODO or XXX, depending on your level of confidence that the code is buggy.

I think we should make the final decision _after_ addressing your second concern discussed below:

> Another question is [whether to] remove the new flag and use the old one instead

We should do that if the flags should represent the same concept. Otherwise, we should rename our new flag.

Please summarize the difference in how HttpRequest::internal and HttpRequest::flags::internal are defined/set/used today. If the two flags are nearly identical (bugs notwithstanding), then I am a bit surprised that you have not stumbled upon the existence of the old flag while adding all the code the sets and uses the new flag...

Comment (eduard-bagdasaryan) [reply src/tunnel.cc]: > Please summarize the difference in how HttpRequest::internal and HttpRequest::flags::internal are defined/set/used today. 

* HttpRequest::internal is assigned by Squid instance itself when it creates HttpRequest object to send it by means of downloader or esi.
*  `HttpRequest::flags::internal` is assigned after receiving a request from another Squid instance if internalCheck() check returns true (the parsed path matched some hardcoded string in internal.cc).

(1) makes setNoAddr() call, so that the client ip will be logged as "-".
(2) does not make setNoAddr(), which is probably reasonable: we want to be able to log the client ip of the received 'internal' request.

It is not clear for me why Squid does not call setNoAddr() after creating 'internal' requests in contexts different from downloader and esi (e.g., peer_digest). Should not client ips for such requests be logged as "-" either? Probably these other requests (peer_digest, etc.) are not associated with ALE, but I am not sure.

File: src/auth/AclProxyAuth.cc [hunk 10]
```diff
@@ -141,14 +141,14 @@ ProxyAuthLookup::LookupDone(void *data)
...
-    if (checklist->auth_user_request == NULL || !checklist->auth_user_request->valid() || checklist->conn() == NULL) {
+    if (!checklist->auth_user_request || !checklist->auth_user_request->valid() || !checklist->clientConnectionManager()) {
         /* credentials could not be checked either way
          * restart the whole process */
         /* OR the connection was closed, there's no way to continue */
         checklist->auth_user_request = NULL;
 
-        if (checklist->conn() != NULL) {
-            checklist->conn()->setAuth(NULL, "proxy_auth ACL failure");
+        if (checklist->clientConnectionManager()) {
```
Comment (rousskov) [inline src/auth/AclProxyAuth.cc]:
```suggestion
        if (const auto mgr = checklist->clientConnectionManager()) {
```

File: src/acl/FilledChecklist.cc [hunk 11]
```diff
@@ -245,17 +252,73 @@ void ACLFilledChecklist::setRequest(HttpRequest *httpRequest)
...
-
-        if (request->clientConnectionManager.valid())
-            conn(request->clientConnectionManager.get());
+        if (Config.onoff.acl_uses_indirect_client) {
+            assert(src_addr != request->indirectClientAddr());
+            src_addr = request->indirectClientAddr();
+        } else
+#endif
+        {
+            assert(src_addr != request->clientAddr());
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/acl/FilledChecklist.cc]: Fixed here but left the old existing assert in ACLFilledChecklist::fd().

File: src/acl/FilledChecklist.cc [hunk 12]
```diff
@@ -245,17 +252,73 @@ void ACLFilledChecklist::setRequest(HttpRequest *httpRequest)
...
-            src_addr = request->indirect_client_addr;
-        else
-#endif /* FOLLOW_X_FORWARDED_FOR */
-            src_addr = request->client_addr;
-        my_addr = request->my_addr;
-
-        if (request->clientConnectionManager.valid())
-            conn(request->clientConnectionManager.get());
+        if (Config.onoff.acl_uses_indirect_client) {
+            assert(src_addr != request->indirectClientAddr());
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/acl/FilledChecklist.cc]: Fixed and used Must instead.

File: src/HttpRequest.h [hunk 13]
```diff
@@ -146,13 +149,38 @@ class HttpRequest: public Http::Message
...
+    CbcPointer<ConnStateData> &clientConnectionManager();
+
+    /// the client connection of the underlying transaction, if any
+    Comm::ConnectionPointer clientConnection() const;
 
 #if FOLLOW_X_FORWARDED_FOR
-    Ip::Address indirect_client_addr;
-#endif /* FOLLOW_X_FORWARDED_FOR */
+    /// the indirect client address
+    const Ip::Address& indirectClientAddr() const;
```
Comment (rousskov) [inline src/HttpRequest.h]: When the indirect client address is unknown (including ignored), a reader would expect indirectClientAddr() to return an empty address or even throw an exception. However, indirectClientAddr() would often return direct client address (i.e. clientAddr()) in that case. The method name and useless description promise an indirect client address but the method sometimes returns a direct one. Sooner or later, a caller will be confused by this substitution. **If** the implementation is correct and all callers should be happy with the direct client address when the indirect one is unavailable, then I suggest renaming this method to furthestClientAddress().

File: src/acl/FilledChecklist.h [hunk 14]
```diff
@@ -38,20 +39,41 @@ class ACLFilledChecklist: public ACLChecklist
...
-    ConnStateData * conn() const;
+    /// Configure client connection-related information. Each and both parameters
+    /// can be nil (or equivalent). Specifying mgr->clientConnection as the second
+    /// parameter is useless.
+    void setClientConnectionDetails(ConnStateData *, Comm::ConnectionPointer conn = nullptr);
+
+#if FOLLOW_X_FORWARDED_FOR
+    /// Configures clientAddr() to always return available indirect client address
+    /// instead of direct client address.
+    void forceIndirectAddr();
```
Comment (rousskov) [inline src/acl/FilledChecklist.h]: What happens when indirect client address is not available and forceIndirectAddr() is called?

* If clientAddr() returns a direct client address in this case, then let's rename this method preferIndirectAddr() and adjust the description accordingly.
* Otherwise, let's document that clientAddr() will return an empty address when forced indirect address is not available.

File: src/client_side_request.cc [hunk 15]
```diff
@@ -1702,6 +1684,16 @@ ClientHttpRequest::clearRequest()
...
 }
 
+const Ip::Address&
+ClientHttpRequest::clientAddrOnError() const
+{
+    if (const auto conn = getConn())
+        if (conn->clientConnection)
+            return conn->clientConnection->remote;
+    static Ip::Address emptyAddr;
+    return emptyAddr;
```
Comment (rousskov) [inline src/client_side_request.cc]: We are duplicating these two lines in several places. Let's add an inlined Ip::Address::Empty() instead.

File: src/client_side_request.h [hunk 16]
```diff
@@ -124,10 +127,10 @@ class ClientHttpRequest
     AccessLogEntry::Pointer al; ///< access.log entry
 
     struct Flags {
-        Flags() : accel(false), internal(false), done_copying(false), purging(false) {}
+        Flags() : accel(false), internalReceived(false), done_copying(false), purging(false) {}
 
         bool accel;
-        bool internal;
+        bool internalReceived;
```
Comment (rousskov) [inline src/client_side_request.h]:
```suggestion
        /// a request with a /squid-internal-... URL referring to this Squid instance
        bool askingForOurInternalResource;
```

File: src/HttpRequest.h [hunk 17]
```diff
@@ -146,13 +149,38 @@ class HttpRequest: public Http::Message
 
     int imslen;
 
-    Ip::Address client_addr;
+    /// mark this request as initiated by Squid (rather than received on a client connection)
+    void selfInitiated();
```
Comment (rousskov) [inline src/HttpRequest.h]: We need to rename this so that it does not look like a test method. I cannot think of a good name, but this option is better than leaving it "as is" and better than makeSelfInitiated() IMO:

```suggestion
    void markAsSelfInitiated();
```

File: src/HttpRequest.h [hunk 18]
```diff
@@ -146,13 +149,38 @@ class HttpRequest: public Http::Message
...
+
+    /// whether this request was initiated by Squid (rather than received on a client connection)
+    bool isSelfInitiated() const { return selfInitiated_; }
```
Comment (rousskov) [inline src/HttpRequest.h]: After selfInitiated() above is fixed:

```suggestion
    bool selfInitiated() const { return selfInitiated_; }
```

**Comment by eduard-bagdasaryan** [comment]
I found three places where self-initiated requests are used:
* Downloader
* Esi
* Asn

> Are all self-initiated requests logged in access.log (by default)? None are logged? Some?

The logging is performed when ClientHttpRequest's destructor calls logRequest(). I see that Downloader and Esi code create this object (so the logging will be performed), but Asn - does not (no logging).

> Does each self-initiated request get a dedicated ALE object?

ALE is created in ClientHttpRequest constructor. Asn does not create it separately. So, again, the first two cases get a dedicated ALE, while the third does not.

> Do all self-initiated requests triggered by the same master transaction share the corresponding MasterXaction object? I understand that not all self-initiated requests are triggered by some master transaction. For this question, let's consider only self-initiated requests that were triggered by or associated with some master transaction.

In all cases, a new MasterXaction object is created. This object is passed then into the HttpRequest constructor.

File: src/HttpRequest.cc [hunk 19]
```diff
@@ -734,34 +733,82 @@ UpdateRequestNotes(ConnStateData *csd, HttpRequest &request, NotePairs const &he
...
 
-    AnyP::PortCfgPointer port = clientConnectionManager->port;
-    if (port) {
-        myportname = port->name;
-        flags.ignoreCc = port->ignore_cc;
-    }
+const Ip::Address&
+HttpRequest::clientAddr() const
+{
+    if (!selfInitiated_)
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/HttpRequest.cc]: Done.

File: src/HttpRequest.cc [hunk 20]
```diff
@@ -822,9 +875,18 @@ FindListeningPortAddress(const HttpRequest *callerRequest, const AccessLogEntry
         return ip;
 
     /* handle non-intercepted cases that were not handled above */
-    ip = FindListeningPortAddressInConn(request->masterXaction->tcpClient);
+    ip = FindListeningPortAddressInConn(request->clientConnection());
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/HttpRequest.cc]: I checked that the old code did not initialize masterXaction->tcpClient field for self-initiated (Downloader, Asn, Esi) requests.

File: src/HttpRequest.h (hunk 17 above)
Comment (eduard-bagdasaryan) [reply src/HttpRequest.h]: Done.

File: src/HttpRequest.h (hunk 18 above)
Comment (eduard-bagdasaryan) [reply src/HttpRequest.h]: Done.

File: src/HttpRequest.h [hunk 21]
```diff
@@ -146,13 +149,38 @@ class HttpRequest: public Http::Message
...
+
+    /// supply Downloader-specific settings
+    void prepareForDownloader(Downloader *);
+
+    /// specify addresses manually when lacking client connection
+    void prepareForConnectionlessProtocol(const Ip::Address &fromAddr, const Ip::Address &localAddr);
+
+    /// the remote address of the client connection
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/HttpRequest.h]: Partially done: clientAddr() is about derect addresses only, while furthestClientAddress() may return direct or indirect.

File: src/HttpRequest.h [hunk 22]
```diff
@@ -146,13 +149,38 @@ class HttpRequest: public Http::Message
...
 
 #if FOLLOW_X_FORWARDED_FOR
-    Ip::Address indirect_client_addr;
-#endif /* FOLLOW_X_FORWARDED_FOR */
+    /// Indirect client address, if available, otherwise clientAddr().
+    const Ip::Address& furthestClientAddress() const;
 
-    Ip::Address my_addr;
+    void indirectClientAddr(const Ip::Address &addr) { indirect_client_addr = addr; }
+    /// force using direct client address
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/HttpRequest.h]: Done with mentioning the only furthestClientAddress().

File: src/HttpRequest.h [hunk 23]
```diff
@@ -146,13 +149,38 @@ class HttpRequest: public Http::Message
...
+    const Ip::Address& clientAddr() const;
+
+    /// the local address of the client connection
+    const Ip::Address& myAddr() const;
+
+    /// the client connection manager of the underlying transaction, if any
+    CbcPointer<ConnStateData> &clientConnectionManager();
+
+    /// the client connection of the underlying transaction, if any
+    Comm::ConnectionPointer clientConnection() const;
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/HttpRequest.h]: Done. However, since MasterXaction::clientConnection_ should be nil for self-initiated requests, I think we should change the implementation, returning simply masterXaction->clientConnection().

File: src/acl/FilledChecklist.cc [hunk 24]
```diff
@@ -245,17 +254,78 @@ void ACLFilledChecklist::setRequest(HttpRequest *httpRequest)
     if (httpRequest) {
         request = httpRequest;
         HTTPMSGLOCK(request);
+        setClientConnectionDetails(request->clientConnectionManager().get());
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/acl/FilledChecklist.cc]: I think we should ignore because we do not have client connection for self-initiated requests.

File: src/acl/FilledChecklist.cc [hunk 25]
```diff
@@ -245,17 +254,78 @@ void ACLFilledChecklist::setRequest(HttpRequest *httpRequest)
...
+        if (!clientConnectionManager()) // could not take the connection from the connection manager
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/acl/FilledChecklist.cc]: I think this handles the case when MasterXaction::clientConnectionManager_ is nil but MasterXaction::clientConnection_ is not. Such MasterXaction is created in TcpAcceptor::notify().

File: src/comm.cc [hunk 26]
```diff
@@ -468,7 +468,7 @@ comm_apply_flags(int new_socket,
     if ( (flags & COMM_DOBIND) || addr.port() > 0 || !addr.isAnyAddr() ) {
         if ( !(flags & COMM_DOBIND) && addr.isAnyAddr() )
             debugs(5, DBG_IMPORTANT,"WARNING: Squid is attempting to bind() port " << addr << " without being a listener.");
-        if ( addr.isNoAddr() )
+        if (!addr.isBindable())
```
Comment (eduard-bagdasaryan) [inline src/comm.cc]: This change caused a problem: the new condition evaluates to 'true' for unspecified (i.e., 'any') address and the debug message below is displayed. This logic seems correct: we report about 'not-bindable' address. The isBindable(), which is equivalent to !isEmpty() && !isNoAddr() seems correct either. There problem here, IMO is that Ip::Address cannot distinguish two states: isEmpty() and isAnyAddr(). For example, if we create an empty Ip::Address and call setAnyAddr() on it, isEmtpy() still returns true. I suspect that the roots of this problem come from the fact that IN_ADDR_ANY is usually 0 and leaving zero sin_addr/sin6_addr when initializing sockaddr_in/sockaddr_in6 structures also means 'bind to any address'. So, passing a brand new Ip::Address to bind() has the same effect as calling setAnyAddr() and bind(). I wonder whether we need some new flag in Ip::Address  to distinguish these two states.

Comment (rousskov) [reply src/comm.cc]: If isAnyAddr() usually implies isEmpty() and isEmpty() usually implies isAnyAddr(), but the calling code assumes that those are always two different states, than something went wrong. Moreover, isAnyAddr() documentation already says (badly) that "any address" is the initial state of an Ip::Address object. We should not be adding isEmpty() which claims to test the same thing.

You have added isEmpty() for some specific reason. What was that reason? I did not find that reason in commit 1d9260b. And why do we need isEmpty() today?

BTW, commit 1d9260b appears to be changing the meaning of the HttpRequest::indirectClientAddr() code because `IpAddress::operator==` used by isEmpty() ignores port differences, and the caller looked at the port before that commit. I did not trace what happened to that calling code; perhaps it is already fixed. It was the only isEmpty() user at the commit 1d9260b time.

I do not know whether it is important or not right now, but it sounds like isBindable() simply means !noAddr() -- we can call bind() with all other addresses AFAICT.

Commit a3315b8 talks about distinguishing the three special addresses, but it appears to be based on the false assumption that there are three special addresses (instead of two), and it is difficult for me to extract useful information (such as meaningful examples that would justify isEmtpy() existence) from that commit message.

> I wonder whether we need some new flag in Ip::Address to distinguish these two states.

Before we discuss the (very remote!) possibility of adding flags to Ip::Address, let's clarify: Why do we need (or "What code needs...") to distinguish the following two Ip::Address states?

1. an Ip::Address carrying `IN_ADDR_ANY` IP address and any port value
2. default-created Ip::Address

If that is not the distinction we are after, then please state what problem we are trying to solve by adding isEmpty().

Comment (eduard-bagdasaryan) [reply src/comm.cc]: I think we need to consider Ip::Address use cases, before deciding whether isEmpty() (and isAnyAddr()) is necessary/applicable/different_meaning etc.  In general, I assume that 'any addr' (and underlying INADDR_ANY socket option) may be specified for server (bindable) socket only.

1. Local (listening server) address, passed to bind(). isEmpty() is not needed here and isAnyAddr() works well instead.
   isBindable() now returns just "!isNoAddr()" thus fixing the underlying problem.

2. Local(client) addresses used in self-generated requests (Downloader, etc). isAnyAddr() has no sense for clients: we probably need to call isEmtpy() (though the both methods returning true correspond to the same Ip::Adress state). Note that before the branch changes such addresses were configured as 'no addr'.

3. Remote client/server addresses. Same as (2): if these addresses cannot be acquired, I think we need to use isEmpty() API, emphasizing that the Ip::Address is still in its pristine state. Before the branch changes some (or may be all) contexts used 'no addr' if the remote address was unknown (for example, when responding with error page).

So, in short, I think that isAnyAddr() is applicable to the local(server) addresses, while isEmtpy() - to all other addresses. If I am right, we probably need some indication in Ip::Address about this kind, so that we were obliged to call a correct method in each context. For now, we can leave just a comment about that.

Comment (eduard-bagdasaryan) [reply src/comm.cc]: As we discussed, it would be better to eliminate isEmpty() method. At the same time, we still need an indication (other from the inapplicable isAnyAddr()) of whether client address 'is empty', i.e., still has a default value. After some reflection, I dismissed the approach we discussed previously (due to implementation complexity) of creating a 'proxy' class for Ip::Address, which would keep the information about such 'emptiness'.  Instead, I used the isKnown() method in the two affected contexts (37bd940b). This method is used already in many other contexts in this branch (and probably should have been used in these two contexts also).
Basically, we have two 'higher level' Address validity checking methods: isBindable() and isKnown(). The first one should be used for checking (1) [type](https://github.com/measurement-factory/squid/pull/20/files#r354874746) while the second - for (2) and ( 3) types.
